File: rust/src/filter.rs
use std::collections::HashMap;
use sea_orm::{sea_query::{Alias, Expr, PostgresQueryBuilder, Query}, DatabaseConnection, DbBackend, DbErr, EntityTrait, Iterable, Order, Statement};
use sea_orm::sea_query::extension::postgres::PgExpr;
use sea_orm::sea_query::{Func, SimpleExpr};
use crate::dto::{WithTotalTrait, GridFilter, PaginatedResponse};

macro_rules! unaccent {
    ($col:expr) => {
        SimpleExpr::from(Func::cust(Alias::new("unaccent")).arg($col))
    }
}

pub async fn get_entities_with_total<T: EntityTrait, U: WithTotalTrait>(
    filter: &GridFilter,
    global_searchable_fields: &Vec<T::Column>,
    column_by_name: &HashMap<String, T::Column>,
    db: &DatabaseConnection,
) -> Result<PaginatedResponse<U>, DbErr> {
    let mut query = Query::select();

    let mut query = query
        .columns(T::Column::iter())
        .expr_as(
            Expr::cust("COUNT(*) OVER()"),
            Alias::new("total"),
        )
        .from(T::default());

    // Filter (ignore case)
    for (name, filter) in &filter.filter {
        let column_name = column_by_name.get(name).ok_or(DbErr::Custom("invalid column name".to_string()))?;
        let column = Expr::col(*column_name);
        let filter_no_accent = unidecode::unidecode(&filter.filter);

        match filter.operator.as_str() {
            "equals" => query = query.and_where(unaccent!(column).eq(filter_no_accent)),
            "notEquals" => query = query.and_where(unaccent!(column).ne(filter_no_accent)),
            "contains" => query = query.and_where(unaccent!(column).ilike(format!("%{}%", filter_no_accent))),
            "notContains" => query = query.and_where(unaccent!(column).ilike(format!("%{}%", filter_no_accent)).not()),
            "startsWith" => query = query.and_where(unaccent!(column).ilike(format!("{}%", filter_no_accent))),
            "endsWith" => query = query.and_where(unaccent!(column).ilike(format!("%{}", filter_no_accent))),
            "blank" => query = query.and_where(column.is_null()),
            "notBlank" => query = query.and_where(column.is_not_null()),
            _ => return Err(DbErr::Custom(format!(
                "Unsupported operator: {}.",
                filter.operator
            ))),
        }

    }

    // Global filter (ignore case)
    if !filter.global_search.is_empty() {
        let filter_no_accent = unidecode::unidecode(&filter.global_search);
        let mut or_condition = unaccent!(Expr::col(global_searchable_fields[0])).ilike(format!("%{}%", filter_no_accent));

        for field in &global_searchable_fields[1..] {
            or_condition = or_condition.or(unaccent!(Expr::col(*field)).ilike(format!("%{}%", filter_no_accent)));
        }

        query = query.and_where(or_condition);
    }

    // Sort
    if filter.sort.len() > 0 {
        let sort = &filter.sort.first().unwrap();
        let column = column_by_name.get(&sort.col_id).ok_or(DbErr::Custom("invalid column name".to_string()))?;

        query = query.order_by(
            *column,
            if sort.sort.to_lowercase() == "asc" { Order::Asc } else { Order::Desc },
        );
    } else {
        let column = column_by_name.get("id").ok_or(DbErr::Custom("id column not defined".to_string()))?;
        query = query.order_by(*column,Order::Asc);
    }

    // Paging
    if filter.end != -1 {
        query = query.offset(filter.start as u64).limit((filter.end - filter.start) as u64);
    }

    // Execute query
    let (sql, values) = query.build(PostgresQueryBuilder);

    // For debugging: print the query
    // println!("{}", sql);
    // println!("{:?}", values);

    let rows: Vec<U> = U::find_by_statement(Statement::from_sql_and_values(
        DbBackend::Postgres,
        &sql,
        values,
    ))
        .all(db)
        .await?;

    let total = rows.first().map(|row| row.total()).unwrap_or(0);

    Ok(PaginatedResponse::new(rows, total))
}

File: python-django/pythontest/utils/filter.py
from django.db.models import Q, Window, Count, QuerySet
from typing import List
from unidecode import unidecode

def grid_filter(qs: QuerySet, query, global_searchable_fields: List[str]):
    """Returns the list en entities, filtered and ordered according to the query parameters"""

    qs = qs.annotate(
        total=Window(
            expression=Count('*'),
            partition_by=None
        )
    )

    # Filter
    for column, field_filter in query['filter'].items():
        filter_value = unidecode(field_filter['filter'])
        unaccent_column = f"{column}__unaccent"

        match field_filter['type']:
            case 'equals':
                qs = qs.filter(**{f"{unaccent_column}": filter_value})
            case 'notEquals':
                qs = qs.filter(~Q(**{f"{unaccent_column}": filter_value}))
            case 'contains':
                qs = qs.filter(**{f"{unaccent_column}__icontains": filter_value})
            case 'notContains':
                qs = qs.filter(~Q(**{f"{unaccent_column}__icontains": filter_value}))
            case 'startsWith':
                qs = qs.filter(**{f"{unaccent_column}__istartswith": filter_value})
            case 'endsWith':
                qs = qs.filter(**{f"{unaccent_column}__iendswith": filter_value})
            case 'blank':
                qs = qs.filter(
                    Q(**{f"{column}": ""}) | Q(**{f"{column}__isnull": True})
                )
            case 'notBlank':
                qs = qs.filter(
                    ~(Q(**{f"{column}": ""}) | Q(**{f"{column}__isnull": True}))
                )

    # Global filter
    global_search = query['globalSearch']

    if global_search:
        global_search_normalized = unidecode(global_search)
        q = Q()

        for column in global_searchable_fields:
            unaccent_column = f"{column}__unaccent"
            q = q | Q(**{f'{unaccent_column}__icontains': global_search_normalized})

        qs = qs.filter(q)

    # Sort
    sort = query['sort']

    if sort:
        first_sort = sort[0]

        match first_sort['sort'].lower():
            case 'desc':
                qs = qs.order_by(first_sort['colId'])
            case 'asc':
                qs = qs.order_by(f'-{first_sort["colId"]}')
    else:
        qs = qs.order_by('id')

    qs = qs[query['start']:query['end']]

    data = list(qs)
    total = data[0].total if len(data) > 0 else 0

    return {
        "data": data,
        "total": total,
    }
